### PR TITLE
Arbitrary bytes in blob store register

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today the blob store register supports recording only a `long`, represented as an 8-byte blob. We need to store a little more data in the register, so this commit generalises things to work with a `BytesReference` directly.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/blobstore/support/BlobContainerUtils.java [hunk 1]
```diff
@@ -10,35 +10,42 @@
...
 import java.io.IOException;
 import java.io.InputStream;
-import java.nio.ByteBuffer;
 
 public class BlobContainerUtils {
     private BlobContainerUtils() {
         // no instances
     }
 
+    public static final int MAX_REGISTER_CONTENT_LENGTH = Long.BYTES;
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/common/blobstore/support/BlobContainerUtils.java:24]: I think this should be greater than `Long.BYTES`?

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 2]
```diff
@@ -798,6 +800,17 @@ public static byte[] randomByteArrayOfLength(int size) {
...
 
+    public static BytesReference randomBytesReference(int length) {
+        final var slices = new ArrayList<BytesReference>();
+        var remaining = length;
+        while (remaining > 0) {
+            final var sliceLen = between(1, remaining);
+            slices.add(new BytesArray(randomByteArrayOfLength(sliceLen)));
+            remaining -= sliceLen;
+        }
+        return CompositeBytesReference.of(slices.toArray(BytesReference[]::new));
```
Comment (fcofdez) [inline test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:811]: Maybe we should generate the other `BytesReference` variants too?

File: x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/RegisterAnalyzeAction.java [hunk 3]
```diff
@@ -250,4 +260,32 @@ public String getDescription() {
             );
         }
     }
+
+    static long longFromBytes(BytesReference bytesReference) {
+        if (bytesReference.length() == 0) {
+            return 0L;
+        } else if (bytesReference.length() == Long.BYTES) {
+            try (var baos = new ByteArrayOutputStream(Long.BYTES)) {
```
Comment (fcofdez) [inline x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/RegisterAnalyzeAction.java:268]: Don't we have some library functions to deal with this directly?

**Review by fcofdez — COMMENTED** [review]
Looks good, but I left a question regarding `MAX_REGISTER_CONTENT_LENGTH`

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java (hunk 2 above)
Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:811]: IMO this is enough - it means that we make `#iterator()` yield a fairly random selection of slice sizes, which is where I suspect the most bugs lurk. We'll also occasionally return a `BytesArray` in the case that there's only one slice.

File: x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/RegisterAnalyzeAction.java (hunk 3 above)
Comment (DaveCTurner) [reply x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/RegisterAnalyzeAction.java:268]: Apparently not, or at least I couldn't find any.

**Review by fcofdez — APPROVED** [review]
LGTM, thanks for the clarifications 👍

**Comment by DaveCTurner** [comment]
Well this is frustrating. https://gradle-enterprise.elastic.co/s/jd6eqec3nbvfe looks like a related failure, but I've been running that test in a loop for ages without being able to reproduce it.

**Comment by DaveCTurner** [comment]
I have been running this test in a loop for 20+ hours without another failure. I'm going to ignore it in the name of progress.

**Pull request merged**